ethsigner/core/src/main/java/tech/pegasys/ethsigner/core/requesthandler/AbstractRequestHandler.java

```diff
@@ -0,0 +1,93 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.ethsigner.core.requesthandler;
+
+import static io.netty.handler.codec.http.HttpResponseStatus.GATEWAY_TIMEOUT;
+import static io.netty.handler.codec.http.HttpResponseStatus.INTERNAL_SERVER_ERROR;
+
+import java.net.ConnectException;
+import java.time.Duration;
+import java.util.concurrent.TimeoutException;
+
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.HttpClientRequest;
+import io.vertx.core.http.HttpClientResponse;
+import io.vertx.ext.web.RoutingContext;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public abstract class AbstractRequestHandler {
+
+  private static final Logger LOG = LogManager.getLogger();
+  private final Duration httpRequestTimeout;
+
+  public AbstractRequestHandler(Duration httpRequestTimeout) {
+    this.httpRequestTimeout = httpRequestTimeout;
+  }
+
+  private void handleException(final RoutingContext context, final Throwable thrown) {
+    if (thrown instanceof TimeoutException || thrown instanceof ConnectException) {
+      context.fail(GATEWAY_TIMEOUT.code(), thrown);
+    } else {
+      context.fail(INTERNAL_SERVER_ERROR.code(), thrown);
+    }
+  }
+
+  protected void handleResponse(final RoutingContext context, final HttpClientResponse response) {
+    logResponse(response);
+
+    response.bodyHandler(
+        body -> {
+          LOG.info("Handling received body");
+          context
+              .vertx()
+              .executeBlocking(
+                  future -> {
+                    logResponseBody(body);
+                    handleResponseBody(context, response, body);
+                    future.complete();
+                  },
+                  false,
+                  (res) -> {
+                    if (res.failed()) {
+                      LOG.error(
+                          "An unhandled error occurred while processing {}",
+                          context.getBodyAsString(),
+                          res.cause());
+                      context.fail(res.cause());
+                    }
+                  });
+        });
+  }
+
+  protected void sendRequest(
+      final HttpClientRequest request, final Buffer bodyContent, final RoutingContext context) {
+    request.setTimeout(httpRequestTimeout.toMillis());
+    request.exceptionHandler(thrown -> handleException(context, thrown));
+    request.headers().setAll(context.request().headers());
+    request.headers().remove("Content-Length"); // created during 'end'.
+    request.setChunked(false);
+    request.end(bodyContent);
+  }
+
+  protected abstract void handleResponseBody(
```

Comment:
This and the log methods could be private

Reply:
handleResponseBody needs to be protected such that it can be overridden - logs however +1.